### PR TITLE
Kconfig support for stm32 HAL/LL file selection

### PR DESCRIPTION
**Pull request by lowlander**
These commits add the possibility to select parts of the STM32 CUBE HAL/LL by simply adding a USE_STM32_xxXX_HAL_*  entry to a Kconfig file. That way driver writers do not have to change the Kconfig/Kbuild files under ext/hal/. Also users can use the HAL without needing to change Zephyr code.

### CODE REVIEW

**Comment by erwango** [comment]
Understand what you're trying to do, but I have 3 concerns:
-moving build logic in drivers Kconfig files
-creation of huge number of Kconfig flags (compile duration impact?)
-This is making huge files difficult to review even if there should be much less impact on these files afterwards.

Is there a way to split these files and distribute it in ext/hal/st/stm32cube/stm32yyxx/ folders?
Then only dedicated part of the flags/files would be visible when compiling.

One last point is that LL headers could be used without any flags (as done for serial driver case for instance). Maybe some comment somewhere should make this clear (USE_LL flags are only valid for *_LL_*.c files). This is not new, but creation of USE_LL flags might be misleading in that regard.

**Comment by lowlander** [comment]
@erwango 

> -moving build logic in drivers Kconfig files

Is that really worse than having driver build logic in a completely different part of the Zephyr source tree like it is now? With this change the logic is all local to the driver.  Also it will cause less commit "collisions" cause at the moment all HAL/LL users will need to edit the etx/hal/st/ kbuild files.  

Also when more than one driver (for example PWM and Counter) need the same hal files (tim.c) the current kbuild file will get more and more cluttered. If there are than also user applications, demo's or tests that need HAL/LL sources it will get even worse. 

Another thing is now the hal kbuid file has things like;

`obj-$(CONFIG_SERIAL_HAS_DRIVER) += stm32f0xx/drivers/src/stm32f0xx_hal_uart.o`

That doesn't give any information which serial driver actually needs the HAL. And if there is a serial driver that doesn't need the HAL it can not disable compiling of the HAL code.  Same for CONFIG_PWM and CONFIG_I2C.

> -creation of huge number of Kconfig flags (compile duration impact?)

The "if SOC_SERIES_STM32F4X" already reduces that it seems. Splitting it up in one file per SOC-type will make the file more readable.

> -This is making huge files difficult to review even if there should be much less impact on these files afterwards.

As long as no files are added or removed from the Cube HAL the kconfig/kbuild file should never be touched anymore.  

Maybe one could use some python script to auto generate the kbuild and kconfig files ? They are pretty much one to one with the *.c files.  If it is hard to do that at build time, the cube maintainer could use the script to generate the files at "commit" time?

> One last point is that LL headers could be used without any flags (as done for serial driver case for instance). Maybe some comment somewhere should make this clear (USE_LL flags are only valid for LL.c files). This is not new, but creation of USE_LL flags might be misleading in that regard.

Maybe a different verb than USE?  Something like LINK_TO__STM32F1XX_LL_I2C  or NEED_STM32F1XX_LL_I2C _OBJ ?

**Comment by erwango** [comment]
> Is that really worse than having driver build logic in a completely different part of the Zephyr source tree like it is now? With this change the logic is all local to the driver. Also it will cause less commit "collisions" cause at the moment all HAL/LL users will need to edit the etx/hal/st/ kbuild files.

> Also when more than one driver (for example PWM and Counter) need the same hal files (tim.c) the current kbuild file will get more and more cluttered. If there are than also user applications, demo's or tests that need HAL/LL sources it will get even worse.

Ok, this is a fair point.

>>    -creation of huge number of Kconfig flags (compile duration impact?)

>The "if SOC_SERIES_STM32F4X" already reduces that it seems. Splitting it up in one file per SOC-type will make the file more readable.

Splitting up files would be nice

>>    -This is making huge files difficult to review even if there should be much less impact on these files afterwards.

> As long as no files are added or removed from the Cube HAL the kconfig/kbuild file should never be touched anymore.

> Maybe one could use some python script to auto generate the kbuild and kconfig files ? They are pretty much one to one with the *.c files. If it is hard to do that at build time, the cube maintainer could use the script to generate the files at "commit" time?

Doing it at build time would require pretty good documentation, to make it understandable for user.
Let keep it static for now. Splitting it up would allow easier generation at 'commit' time.

>>    One last point is that LL headers could be used without any flags (as done for serial driver case for instance). Maybe some comment somewhere should make this clear (USE_LL flags are only valid for LL.c files). This is not new, but creation of USE_LL flags might be misleading in that regard.

> Maybe a different verb than USE? Something like LINK_TO__STM32F1XX_LL_I2C or NEED_STM32F1XX_LL_I2C _OBJ ?

USE_STM32F1XX_LL_I2C _OBJ ? Not quite sure here..

**Comment by lowlander** [comment]
@erwango I split the Kconfig and Kbuild files up and put them in the cpu dirs, that really cleans up the code.  

To the name LINK_TO_STM32F1XX_LL_I2C, NEED_STM32F1XX_LL_I2C_OBJ, USE_STM32F1XX_LL_I2C, etc. I really don't know.  Maybe someone else can comment on this ? 
The most difficult part of software engineering, what should be the name of my class, function, variable :-)

**Comment by lowlander** [comment]
@galak @erwango any ideas what to do with the Kconfig menu entry names ?  Should I change it to USE_STM32F1XX_LL_I2C_OBJ ? And anything else I can do to get this patch accepted.

@dbkinder there are README's in the st hal directories. But those don't end up in the official documentation, do they ? The maintainer notes will probably be of no use to normal users, but the how to use the HAL is of use to them. What would be the way to get that in the "official" documentation?  (assuming this patch is accepted of course)

**Comment by dbkinder** [comment]
The documentation gathering process only looks for RestructuredText-formatted .rst files, so no, the .txt files are ignored when creating our public documentation. If you separate the maintainer notes (in the .txt file) from the public/user notes (in a .rst file) we can get them included.

**Comment by lowlander** [comment]
@dbkinder  thanks for the info, if I have a go from the code maintainers I will split up the documentation and make a rst file out of it.

**Comment by erwango** [comment]
@lowlander : Let's keep USE_STM32YYXX_LL_FOO and document the fact that using LL headers don't require any flags activation.

As a next step, would be good to get feedback from other frequent and active STM32 contributors like @mbolivar, @ydamigos , @superna9999

File: ext/hal/Kbuild
Comment (superna9999) [inline ext/hal/Kbuild]: This header seems inaccurate !

**Comment by superna9999** [comment]
I'm ok with @erwango about the fact we can include headers without necessarily building the *.c files

Except the unnecessary header on ext/hal/Kbuild I'm ok with this.

File: ext/hal/Kbuild
Comment (lowlander) [reply ext/hal/Kbuild]: Probably just added headers to all open files in my editor that didn't have a header asuming they were new files. 

Remove or change ?

File: drivers/serial/Kconfig.stm32
Comment (ydamigos) [inline drivers/serial/Kconfig.stm32]: Does the STM32F0 family use the HAL serial driver? I thought that all families use the LL serial api.

**Review by ydamigos — COMMENTED** [review]
I am fine with the changes. Just a comment which probably is not relevant with this PR.

File: drivers/serial/Kconfig.stm32
Comment (superna9999) [reply drivers/serial/Kconfig.stm32]: F0 should use the same as all STM32 families

File: ext/hal/Kbuild
Comment (superna9999) [reply ext/hal/Kbuild]: Remove please

File: drivers/serial/Kconfig.stm32
Comment (ydamigos) [reply drivers/serial/Kconfig.stm32]: Could we remove this line or should we do it in a separate PR?

Comment (erwango) [reply drivers/serial/Kconfig.stm32]: Should be a conversion of an oversight from F0 PR merged following merge of the PR that converted serial to full LL (Not quite clear, but you got my point).
To do it in a clean way, I propose an initial commit that remove culprit line in ext/hal/st/stm32cube/Kbuild.

Comment (superna9999) [reply drivers/serial/Kconfig.stm32]: Acked

Comment (lowlander) [reply drivers/serial/Kconfig.stm32]: I modeled the driver-kconfig files after the hal-Kconfig file that had;
`obj-$(CONFIG_SERIAL_HAS_DRIVER) += stm32f0xx/drivers/src/stm32f0xx_hal_uart.o`

File: ext/hal/Kbuild
Comment (lowlander) [reply ext/hal/Kbuild]: will do.

File: drivers/serial/Kconfig.stm32
Comment (erwango) [reply drivers/serial/Kconfig.stm32]: That was my point this line was an oversight following of F0 series introduction.
Migration from HAL to LL in serial driver has been merged during process of F0 review. It should have been detected at that point (my bad).
It would be nice if you could do a preliminary commit to your PR where you remove this line.

**Comment by mbolivar** [comment]
> As a next step, would be good to get feedback from other frequent and active STM32 contributors like @mbolivar, @ydamigos , @superna9999

I think the overall approach is clever and have no problem with the changes.

It'd be nice to have CMakeFiles.txt equivalents in cmake_migration :wink:

**Comment by erwango** [comment]
ok seems that change has community approval!
What about @galak? Are you ok with this proposal too?

File: drivers/serial/Kconfig.stm32
Comment (erwango) [reply drivers/serial/Kconfig.stm32]: @lowlander : Btw, just observed that initial Kbuild has been fixed with #4493.
So you could remove it safely after rebase

**Comment by lowlander** [comment]
@erwango  yes I noticed  yesterday evening. I wanted to update the docu (readme.txt) before pushing an update to this PR.

**Comment by lowlander** [comment]
We can close this because with cmake everything is different. A warning might have been useful, now it was a big waste of my time.

**Comment by lowlander** [comment]
Well the "damage" done by the move to cmake was less bad than expected (mostly search and replace). But boy was last week a wrong week to have to deal with that ;-)

**Review by superna9999 — APPROVED** [review]

File: ext/hal/st/stm32cube/README
Comment (erwango) [inline ext/hal/st/stm32cube/README]: We could specify LL header files here

File: ext/hal/st/stm32cube/README
Comment (erwango) [inline ext/hal/st/stm32cube/README]: And then add HAL / LL here

File: ext/hal/st/stm32cube/README
Comment (erwango) [inline ext/hal/st/stm32cube/README]: maybe add a second example like:
	select USE_STM32F4XX_LL_UART if SOC_SERIES_STM32F4X

**Review by erwango — CHANGES_REQUESTED** [review]
Minor LL related clarifications in README and we're good!
Thanks

**Comment by erwango** [comment]
@lowlander, after you fix the last LL bit in README, we'll try to get this change approved at upper level.

**Comment by lowlander** [comment]
@erwango Zephyr is a weekend-job for me, so I will get to it tomorrow or on Sunday.

**Comment by erwango** [comment]
@lowlander : Ok, Thanks for your time! Anytime is ok then!

**Comment by lowlander** [comment]
@erwango dunno if you get an automatic msg when the code is updated, but I updated the read me file.

**Review by erwango — APPROVED** [review]
@lowlander : I didn't received notification for the README update, but I got your message... And then PR is fully ok for me now! Thanks

**Comment by nashif** [comment]
recheck

File: drivers/i2c/Kconfig
Comment (superna9999) [inline drivers/i2c/Kconfig]: The F0x I2C has been merged in the meantime, can you add it ?

**Comment by galak** [comment]
Can we update this with recent merges

**Comment by lowlander** [comment]
@superna9999  F0 I2C Will be in the next push. @galak do you have anything else like the I2C from @superna9999  or were you just revering to the "merge conflict" ?

**Comment by lowlander** [comment]
@galak , @erwango any chance of getting this in master before Xmas ?

**Comment by erwango** [comment]
@galak : Before requesting a rebase to @lowlander, do you have any comments beforehand on this change?

**Comment by galak** [comment]
Can you rebase/update to resolve conflicts.

**Comment by galak** [comment]
I haven't read the full thread, so sorry if this got discussed.  Instead of having:

+ 	select USE_STM32F0XX_LL_UTILS if SOC_SERIES_STM32F0X
+ 	select USE_STM32F1XX_LL_UTILS if SOC_SERIES_STM32F1X
+ 	select USE_STM32F3XX_LL_UTILS if SOC_SERIES_STM32F3X
+ 	select USE_STM32F4XX_LL_UTILS if SOC_SERIES_STM32F4X
+ 	select USE_STM32L4XX_LL_UTILS if SOC_SERIES_STM32L4X
  	help

Can we just do:

select USE_STM32_LL_UTILS

(similar comment for config vars)

**Comment by erwango** [comment]
@lowlander : You need to add a comment to commit "ext: hal: stm32cube: update README". This is a new rule coming from TSC since last week.

Then, last question: Following change proposed by @galak , is there something that prevent to factorize Kconfig from to stm32cube/stm32yyxx/Kconfig to stm32cube/Kconfig  ?
I know we did the opposite move initially, but given recent change, I think this could simplify a step further.

**Comment by lowlander** [comment]
@erwango yeah I noticed that commits with just a title aren't enough anymore,  I am sure others already have discussed the usefulness of that rule.  

To moving the Kconfig files to stm32cube, I think that should be possible. Some CPU's don't have all options, but I don't think that should be a problem.

I'll try to cleanup the patch in the next couple of days.

**Comment by galak** [comment]
@lowlander any update on this patch series?

**Comment by galak** [comment]
recheck

**Comment by lowlander** [comment]
@galak I'll fix the patch tomorow, it is to late now.

**Comment by lowlander** [comment]
@galak  @erwango  now things should be OK I think. There is only one (big) Kconfig for all CPU's and a CMakeLists.txt per CPU. There are a few files that are always compiled when the HAL is selected, for the rest it can be done with "select" statements.

**Review by galak — APPROVED** [review]

**Pull request merged**
